python/pylibvw.cc

```diff
@@ -55,10 +61,64 @@ const size_t pDECISION_SCORES = 8;
 
 void dont_delete_me(void*arg) { }
 
-vw_ptr my_initialize(std::string args)
-{ if (args.find_first_of("--no_stdin") == std::string::npos)
+class PyCppBridge : public RED_PYTHON::ExternalBinding {
+  private:
+    py::object* run_object;
+    void* base_learn;
+
+  public:
+      int random_num = 0;
+      PyCppBridge() {}
+
+      void SetRandomNumber(int n)
+      { random_num = n;
+      }
+
+      void ActualLearn(example* ec)
+      { try
+        {
+          py::object run = *this->run_object;
+
+          boost::shared_ptr<example> temp_ptr(ec, dont_delete_me);
+          py::object temp = py::object(temp_ptr);
+
+          run.attr("__call__")(temp);
+        }
+        catch (...)
+        {
+          // TODO: Properly translate and return Python exception. #2169
+          PyErr_Print();
+          PyErr_Clear();
+          THROW("Exception in 'search_run_fn'");
+        }
+      }
+
+      void SetLearner(void* learner){
+        this->base_learn = learner;
+      }
+
+      void SetupPythonSide(py::object learn_object)
+      { this->run_object = new py::object(learn_object);
+      }
+
+      void CallLearn(example* ec)
+      { reinterpret_cast<VW::LEARNER::single_learner *>(this->base_learn)->learn(*ec);
+      }
+};
+
+vw_ptr my_initialize(std::string args, bool with_reduction = false)
+{ vw* foo;
+  if (args.find_first_of("--no_stdin") == std::string::npos)
     args += " --no_stdin";
-  vw*foo = VW::initialize(args);
+
+  if (with_reduction)
+  { auto ext_binding = std::unique_ptr<RED_PYTHON::ExternalBinding>(new PyCppBridge());
```

Comment:
`VW::make_unique`?

Reply:
didn't use it bc its c++14 and I assumed we need to cover 11? unless im wrong

Reply:
`std::make_unique` is C++14, so we have our own wrapper that implements it and is compatible with C++11, `VW::make_unique` as @olgavrou mentioned.